onnxscript/rewriter/ort_fusions/attention.py

```diff
@@ -29,32 +33,37 @@ def pattern(
         qkv_bias,
         # mask_index,
         past,
-        # attention_bias,
+        bias,
         num_heads,
         # scale,
+        q_mul,
+        k_mul,
+        v_mul,
     ):
-        projected = op.MatMul(input, qkv_weight)
-        # Add bias if present
-        if self._has_input_bias:
-            projected = op.Add(projected, qkv_bias)
+        if self._no_slice:
+            query_BSD = op.MatMul(input, q_mul)
+            key_BSD = op.MatMul(input, k_mul)
+            value_BSD = op.MatMul(input, v_mul)
+        else:
+            projected = op.MatMul(input, qkv_weight)
 
-        # Slice packed Matmul QKV into Q, K, and V
-        # Q, K, and V are of shape (B, S, D)
-        query_BSD = op.Slice(
-            projected,
-            _allow_other_inputs=True,
-            _outputs=["query_mm_sliced"],
-        )
-        key_BSD = op.Slice(
-            projected,
-            _allow_other_inputs=True,
-            _outputs=["key_mm_sliced"],
-        )
-        value_BSD = op.Slice(
-            projected,
-            _allow_other_inputs=True,
-            _outputs=["value_mm_sliced"],
-        )
+            # Slice packed Matmul QKV into Q, K, and V
+            # Q, K, and V are of shape (B, S, D)
+            query_BSD = op.Slice(
+                projected,
+                _allow_other_inputs=True,
```

Comment:
A generic comment: whenever we use this (that is, not check the other input parameters of the Slice op), we may need some suitable checks in the check method to ensure correctness. Typically that would mean checking that the axes, starts, ends, steps are the right value. If we expect "axes" to be a specific value, it may be easier to specify it directly in the pattern. The recently introduced `pattern.ANY_VALUE` will allow you to skip values checked in the checker method, while specifying expected values like `[3]` elsewhere.

Reply:
Modified